packages/ra-core/src/controller/input/useReferenceArrayInputController.ts

```diff
@@ -0,0 +1,145 @@
+ * @property {boolean} loaded has the reference already been loaded
+ * @property {string} referenceBasePath basePath of the reference
+ */
+interface ReferenceArrayInputProps {
+    choices: Record[];
+    error?: any;
+    warning?: any;
+    loading: boolean;
+    loaded: boolean;
+    referenceBasePath: string;
+    setFilter: (filter: any) => void;
+    setPagination: (pagination: Pagination) => void;
+    setSort: (sort: Sort) => void;
+}
+
+interface Option {
+    basePath: string;
+    filter?: any;
+    filterToQuery?: (filter: any) => any;
+    input: FieldInputProps<any, HTMLElement>;
+    options?: any;
+    perPage?: number;
+    record?: Record;
+    reference: string;
+    resource: string;
+    sort?: Sort;
+    source: string;
+}
+
+/**
+ * Prepare data for the ReferenceArrayInput components
+ *
+ * @example
+ *
+ * const { choices, error, loaded, loading, referenceBasePath } = useReferenceArrayInputController({
+ *      basePath: 'resource';
+ *      record: { referenceIds: ['id1', 'id2']};
+ *      reference: 'reference';
+ *      resource: 'resource';
+ *      source: 'referenceIds';
+ * });
+ *
+ * @param {Object} option
+ * @param {boolean} option.allowEmpty do we allow for no referenced record (default to false)
+ * @param {string} option.basePath basepath to current resource
+ * @param {string | false} option.linkType The type of the link toward the referenced record. edit, show of false for no link (default to edit)
+ * @param {Object} option.record The The current resource record
+ * @param {string} option.reference The linked resource name
+ * @param {string} option.resource The current resource name
+ * @param {string} option.source The key of the linked resource identifier
+ *
+ * @return {Object} controllerProps Fetched data and callbacks for the ReferenceArrayInput components
+ */
+const useReferenceArrayInputController = ({
+    basePath,
+    filter: defaultFilter,
+    filterToQuery = searchText => ({ q: searchText }),
+    input,
+    perPage = 25,
+    sort: defaultSort = { field: 'id', order: 'DESC' },
+    options,
+    reference,
+    resource,
+    source,
+}: Option): ReferenceArrayInputProps => {
+    const translate = useTranslate();
+
+    const [pagination, setPagination] = useState({ page: 1, perPage });
+    const [sort, setSort] = useState(defaultSort);
+    const [filter, setFilter] = useState('');
+
+    const { data: referenceRecords, loaded } = useGetMany(
+        reference,
+        input.value || []
+    );
+
+    const finalFilter = useMemo(
+        () => ({
+            ...defaultFilter,
+            ...filterToQuery(filter),
+        }),
+        [defaultFilter, filter, filterToQuery]
+    );
+
+    const { data: matchingReferences } = useGetMatching(
+        reference,
+        pagination,
+        sort,
+        finalFilter,
+        source,
+        resource,
+        options
+    );
+
+    const finalReferenceRecords = referenceRecords.filter(
```

Comment:
We need a comment here, like "filter out not found references - happens when the dataProvider doesn't guarantee referential integrity"